File: shared/src/commonMain/kotlin/com/snabbit/runner/shared/features/gamification/domain/model/NudgePill.kt
package com.snabbit.runner.shared.features.gamification.domain.model

/**
 * Trailing coin / red-card pill for a nudge strip — the KMP port of Dart
 * `lib/models/gamification/nudge_pill.dart`. Derived purely from the coin /
 * red-card counts; there is no `badgeText` in the API.
 */
enum class NudgePillKind { None, Coin, RedCard }

data class NudgePill(
    val kind: NudgePillKind,
    val count: Int,
) {
    val hasPill: Boolean get() = kind != NudgePillKind.None && count > 0

    companion object {
        val None = NudgePill(NudgePillKind.None, 0)
    }
}

/** If [goldCoins] > 0 → coin pill; else if [redCards] > 0 → red-card pill; else none. */
fun deriveNudgePill(goldCoins: Int?, redCards: Int?): NudgePill = when {
    (goldCoins ?: 0) > 0 -> NudgePill(NudgePillKind.Coin, goldCoins!!)
    (redCards ?: 0) > 0 -> NudgePill(NudgePillKind.RedCard, redCards!!)
    else -> NudgePill.None
}

File: lib/models/gamification/nudge_pill.dart
/// Trailing strip/row pill for pre-nudges — derived only from [goldCoins] / [redCards]
/// (LLD §2.0). No `badgeText` in the API.
enum NudgePillKind { none, coin, redCard }

class NudgePillDisplay {
  final NudgePillKind kind;
  final int count;

  const NudgePillDisplay._(this.kind, this.count);

  const NudgePillDisplay.none() : this._(NudgePillKind.none, 0);

  const NudgePillDisplay.coin(int count)
      : this._(NudgePillKind.coin, count);

  const NudgePillDisplay.redCard(int count)
      : this._(NudgePillKind.redCard, count);

  bool get hasPill => kind != NudgePillKind.none && count > 0;
}

/// If [goldCoins] > 0 → coin pill; else if [redCards] > 0 → red-card pill; else none.
NudgePillDisplay deriveNudgePill({int? goldCoins, int? redCards}) {
  if (goldCoins != null && goldCoins > 0) {
    return NudgePillDisplay.coin(goldCoins);
  }
  if (redCards != null && redCards > 0) {
    return NudgePillDisplay.redCard(redCards);
  }
  return const NudgePillDisplay.none();
}
